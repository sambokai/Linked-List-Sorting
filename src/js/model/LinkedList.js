class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}

class LinkedList {
  constructor(value) {
    this.length = 0;
    this.head = null;

    if (value) { this.append(value); }
  }

  append(value) {
    const node = new Node(value);
    let currentNode = this.head;

    if (!currentNode) {
      this.head = node;
      this.length += 1;

      return this;
    }

    while (currentNode.next) {
      currentNode = currentNode.next;
    }

    currentNode.next = node;

    this.length += 1;

    return this;
  }

  prepend(value) {
    const node = new Node(value);
    node.next = this.head;
    this.head = node;
    this.length += 1;
  }

  get(index) {
    let current = this.head;
    let count = 0;

    if (this.length === 0) {
      throw new Error('Empty list');
    }

    if (index < 0 || index > this.length) {
      throw new Error('Invalid index');
    }

    while (count < index) {
      current = current.next;
      count += 1;
    }

    return current;
  }

  remove(index) {
    let current = this.head;

    if (index < 0 || index > this.length) {
      throw new Error('Invalid index');
    }

    if (index === 0) {
      this.head = current.next;
      this.length -= 1;
      return this;
    }

    let count = 0;
    let previousNode;
    let nodeToBeDeleted;

    while (count < index) {
      previousNode = current;
      nodeToBeDeleted = current.next;
      current = current.next;
      count += 1;
    }

    previousNode.next = nodeToBeDeleted.next;
    this.length -= 1;
    return this;
  }

  getIndex(predicate) {
    let current = this.head;
    let count = 0;

    if (this.length === 0) {
      throw new Error('Empty list');
    }

    while (count < this.length) {
      if (predicate(current.value)) return count
      current = current.next;
      count += 1;
    }

    return -1;
  }

  clear() {
    this.head = null;
    this.length = 0;
  }

  filter(predicate) {
    const filtered = new LinkedList();
    // eslint-disable-next-line no-restricted-syntax
    for (const element of [...this].filter(predicate)) {
      filtered.append(element);
    }
    return filtered;
  }

  insertionSort(compare = (a, b) => a > b) {
    const items = [...this];
    for (let i = 0; i < items.length; i += 1) {
      const value = items[i];
      let j;
      for (j = i - 1; j > -1 && compare(items[j], value); j -= 1) {
        items[j + 1] = items[j];
      }
      items[j + 1] = value;
    }
    const sorted = new LinkedList();
    items.forEach(element => sorted.append(element));
    return sorted;
  }

  mergeSort(compare = (a, b) => a > b) {
    const sortedArray = LinkedList.mergeArray([...this], compare);
    const sortedList = new LinkedList();
    sortedArray.forEach(element => sortedList.append(element));
    return sortedList;
  }

  static mergeArray(arr, compare) {
    if (arr.length === 1) {
      return arr;
    }

    const middle = Math.floor(arr.length / 2);
    const left = arr.slice(0, middle);
    const right = arr.slice(middle);

    return this.merge(
      this.mergeArray(left, compare),
      this.mergeArray(right, compare),
      compare,
    );
  }

  static merge(left, right, compare) {
    const result = [];
    let indexLeft = 0;
    let indexRight = 0;

    while (indexLeft < left.length && indexRight < right.length) {
      if (compare(right[indexRight], left[indexLeft])) {
        result.push(left[indexLeft]);
        indexLeft += 1;
      } else {
        result.push(right[indexRight]);
        indexRight += 1;
      }
    }

    return result.concat(left.slice(indexLeft)).concat(right.slice(indexRight));
  }

  * [Symbol.iterator]() {
    let element = this.head;

    while (element) {
      yield element.value;
      element = element.next;
    }
  }
}

export default LinkedList;
