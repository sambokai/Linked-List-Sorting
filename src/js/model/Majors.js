const Majors = Object.freeze({
  COMPUTER_SCIENCE: 'cs',
  MATHEMATICS: 'ma',
  HISTORY: 'hi',
  ECONOMICS: 'ec',
});

export default Majors;
