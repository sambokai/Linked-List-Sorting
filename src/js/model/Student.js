import Majors from "./Majors";

let studentIdCounter = 0;
const incrementId = () => {
  studentIdCounter += 1;
  return studentIdCounter;
};

export const resetId = () => {studentIdCounter = 0;};


class Student {
  constructor(firstName, lastName, universityMajor) {
    if (Object.values(Majors).includes(universityMajor)) {
      this.universityMajor = universityMajor;
    } else {
      throw 'Invalid Major';
    }
    this.firstName = firstName;
    this.lastName = lastName;
    this.id = incrementId();
  }
}

export default Student;
