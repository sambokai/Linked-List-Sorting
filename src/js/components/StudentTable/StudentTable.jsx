import React, { Component } from 'react';
import PropTypes from 'prop-types';

import ReactDataGrid from 'react-data-grid';
import { Toolbar } from 'react-data-grid-addons';

import faker from 'faker';

import LinkedList from '../../model/LinkedList';
import Student from '../../model/Student';
import Majors from '../../model/Majors';

class StudentTable extends Component {
  static randomStudent() {
    return new Student(faker.name.firstName(), faker.name.lastName(), this.randomMajor());
  }

  static randomMajor() {
    const keys = Object.keys(Majors);
    return Majors[keys[Math.floor(keys.length * Math.random())]];
  }

  constructor(props) {
    super(props);

    this.state = {
      students: new LinkedList(),
      columns: [
        {
          key: 'id',
          name: 'ID',
          sortable: true,
          filterable: true,
          width: 100,
        },
        {
          key: 'firstName',
          name: 'First Name',
          sortable: true,
          filterable: true,
        },
        {
          key: 'lastName',
          name: 'Last Name',
          sortable: true,
          filterable: true,
        },
        {
          key: 'universityMajor',
          name: 'Major',
          sortable: true,
          filterable: true,
        },
      ],
      filters: {},
      maxStudentGenerateCount: 100,
    };

    this.studentGenerateCountInput = React.createRef();
  }

  componentWillMount() {
    this.generateRandomStudents(this.props.initialRandomStudentCount);
  }

  onClearFilters = () => {
    this.setState({ filters: {} });
  };


  getCellAction = (column, row) => {
    if (column.key === 'id') {
      return [
        {
          icon: 'fas fa-trash-alt',
          callback: () => { this.deleteStudentById(row.id); },
        },
      ];
    }
    return null;
  };

  handleGridSort = (columnKey, direction) => {
    const compare = (a, b) => {
      switch (direction) {
        case 'ASC':
          return a[columnKey] > b[columnKey];
        case 'DESC':
          return a[columnKey] < b[columnKey];
        default:
          return a.id > b.id;
      }
    };

    this.setState(prevState => ({
      students: prevState.students.insertionSort(compare),
    }));
  };

  handleFilterChange = (filter) => {
    this.setState({
      filters: {
        ...this.state.filters,
        [filter.column.key]: filter.filterTerm,
      },
    });
  };

  handleClearTable = () => {
    const clearedTable = this.state.students.clear();
    this.setState({ students: clearedTable });
  };

  handleGenerateStudents = () => {
    const generateCount = this.studentGenerateCountInput.current.value;
    if (generateCount && generateCount > 0) {
      if (generateCount <= this.state.maxStudentGenerateCount) {
        this.generateRandomStudents(generateCount);
      } else {
        // eslint-disable-next-line no-alert
        alert(`Maximum value of students to be generated is ${this.state.maxStudentGenerateCount}. Please retry.`);
      }
    }
  };

  generateRandomStudents(count) {
    // eslint-disable-next-line prefer-destructuring
    const students = this.state.students;

    for (let i = 0; i < count; i += 1) {
      students.append(StudentTable.randomStudent());
    }

    this.setState({ students });
    return students;
  }

  prependNewStudent = () => {
    const students = this.state.students.prepend(StudentTable.randomStudent());
    this.setState({ students });
  };

  appendNewStudent = () => {
    const students = this.state.students.append(StudentTable.randomStudent());
    this.setState({ students });
  };

  filteredRows = () => {
    const predicate = (student) => {
      // eslint-disable-next-line no-restricted-syntax
      for (const key in this.state.filters) {
        // eslint-disable-next-line no-prototype-builtins
        if (this.state.filters.hasOwnProperty(key) && (student[key] === undefined || this.state.filters[key] !== '')) {
          if (
            (key !== 'id' && !student[key].toString().toLowerCase().includes(this.state.filters[key].toLowerCase()))
              || (key === 'id' && student[key].toString().toLowerCase() !== this.state.filters[key].toLowerCase())
          ) {
            return false;
          }
        }
      }
      return true;
    };

    return this.state.students.filter(predicate);
  };

  rowGetter = i => this.filteredRows().get(i).value;

  deleteStudentById(id) {
    const indexToBeDeleted = this.state.students.getIndex(student => student.id === id);
    const studentsWithoutDeleted = this.state.students.remove(indexToBeDeleted);
    this.setState({ students: studentsWithoutDeleted });
  }

  render() {
    return (
      <div>
        <nav className="navbar navbar-dark bg-dark justify-content-between">
          <a id="navbar-logo" className="navbar-brand" href="/">Linked List & Students</a>
          <form className="form-inline">
            <div className="form-group">
              <input ref={this.studentGenerateCountInput} defaultValue={50} className="form-control" type="number" size={4} min={1} max={this.state.maxStudentGenerateCount} />
              <button
                id="new-note-button"
                onClick={this.handleGenerateStudents}
                type="button"
                className="btn btn-primary ml-2"
              >Generate Students
              </button>
            </div>
            <button
              id="delete-note-button"
              onClick={this.handleClearTable}
              type="button"
              className="btn btn-danger ml-2"
            >Clear Table
            </button>
          </form>
        </nav>
        <ReactDataGrid
          columns={this.state.columns}
          rowGetter={this.rowGetter}
          rowsCount={this.filteredRows().length}
          minHeight={600}
          onGridSort={this.handleGridSort}
          toolbar={(
            <Toolbar enableFilter>
              <button className="btn ml-2" onClick={this.prependNewStudent}>Prepend New Student</button>
              <button className="btn ml-2" onClick={this.appendNewStudent}>Append New Student</button>
              <div id="student-count" className="d-inline-block ml-3">Total Students: {this.state.students.length}</div>
            </Toolbar>
          )}
          onAddFilter={this.handleFilterChange}
          onClearFilters={this.onClearFilters}
          getCellActions={this.getCellAction}
        />
      </div>
    );
  }
}

StudentTable.propTypes = {
  initialRandomStudentCount: PropTypes.number,
};

StudentTable.defaultProps = {
  initialRandomStudentCount: 500,
};

export default StudentTable;
