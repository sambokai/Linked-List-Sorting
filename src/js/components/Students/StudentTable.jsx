import React, { Component } from 'react';
import PropTypes from 'prop-types';

import ReactDataGrid from 'react-data-grid';
import faker from 'faker';

import LinkedList from '../../model/LinkedList';
import Student from '../../model/Student';
import Majors from '../../model/Majors';

class StudentTable extends Component {
  constructor(props) {
    super(props);

    this.state = {
      students: null,
      columns: [
        {
          key: 'id',
          name: 'ID',
          sortable: true,
        },
        {
          key: 'firstName',
          name: 'First Name',
          sortable: true,
        },
        {
          key: 'lastName',
          name: 'Last Name',
          sortable: true,
        },
        {
          key: 'universityMajor',
          name: 'Major',
          sortable: true,
        },
      ],
    };
  }

  componentWillMount() {
    this.generateRandomStudents(this.props.randomStudentCount);
  }

  generateRandomStudents(count) {
    const randomStudents = new LinkedList();

    for (let i = 0; i < count; i += 1) {
      const randomMajor = () => {
        const keys = Object.keys(Majors);
        return Majors[keys[Math.floor(keys.length * Math.random())]];
      };
      const student = new Student(faker.name.firstName(), faker.name.lastName(), randomMajor());
      randomStudents.append(student);
    }

    this.setState({ students: randomStudents });
    return randomStudents;
  }

  handleGridSort = (columnKey, direction) => {
    const compare = (a, b) => {
      switch (direction) {
        case 'ASC':
          return a[columnKey] > b[columnKey];
        case 'DESC':
          return a[columnKey] < b[columnKey];
        default:
          return a.id > b.id;
      }
    };

    this.setState(prevState => ({
      students: prevState.students.insertionSort(compare),
    }));
  };

  rowGetter = i => this.state.students.get(i).value;

  render() {
    return (
      <ReactDataGrid
        columns={this.state.columns}
        rowGetter={this.rowGetter}
        rowsCount={this.state.students.length}
        minHeight={400}
        onGridSort={this.handleGridSort}
      />
    );
  }
}

StudentTable.propTypes = {
  randomStudentCount: PropTypes.number,
};

StudentTable.defaultProps = {
  randomStudentCount: 50,
};

export default StudentTable;
