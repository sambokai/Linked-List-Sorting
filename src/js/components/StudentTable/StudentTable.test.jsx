import React from 'react';

import { shallow } from 'enzyme';

import StudentTable from './StudentTable';
import LinkedList from '../../model/LinkedList';
import Majors from '../../model/Majors';
import Student from '../../model/Student';

describe('<StudentTable/>', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<StudentTable randomStudentCount={50} />);

    expect(wrapper).toMatchSnapshot();
  });

  describe('handleGridSort()', () => {
    const directions = ['ASC', 'DESC', 'NONE'];

    directions.forEach((direction) => {
      it(`should sort the linkedlist of students correctly for direction '${direction}'`, () => {
        const wrapper = shallow(<StudentTable randomStudentCount={0} />);
        const keyToBeSortedBy = 'firstName';
        wrapper.setState({
          students: new LinkedList(new Student('Anton', 'Meyer', Majors.COMPUTER_SCIENCE))
            .append(new Student('Xerxes', 'Smith', Majors.MATHEMATICS))
            .append(new Student('George', 'Michael', Majors.HISTORY)),
        });

        wrapper.instance().handleGridSort(keyToBeSortedBy, direction);

        expect([...wrapper.state('students')]).toMatchSnapshot(keyToBeSortedBy + direction);
      });
    });
  });

  describe('rowGetter()', () => {
    it('should return the student at a specified index in the linked-list', () => {
      const wrapper = shallow(<StudentTable randomStudentCount={0} />);
      const testStudent = new Student('Xerxes', 'Smith', Majors.MATHEMATICS);
      wrapper.setState({
        students: new LinkedList(new Student('Anton', 'Meyer', Majors.COMPUTER_SCIENCE))
          .append(testStudent)
          .append(new Student('George', 'Michael', Majors.HISTORY)),
      });

      const actual = wrapper.instance().rowGetter(1);

      expect(actual).toMatchObject(testStudent);
    });
  });
});

