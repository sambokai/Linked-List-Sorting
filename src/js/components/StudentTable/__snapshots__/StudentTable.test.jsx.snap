// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StudentTable/> handleGridSort() should sort the linkedlist of students correctly for direction 'ASC': firstNameASC 1`] = `
Array [
  Student {
    "firstName": "Anton",
    "id": 1,
    "lastName": "Meyer",
    "universityMajor": "Computer Science",
  },
  Student {
    "firstName": "George",
    "id": 3,
    "lastName": "Michael",
    "universityMajor": "History",
  },
  Student {
    "firstName": "Xerxes",
    "id": 2,
    "lastName": "Smith",
    "universityMajor": "Mathematics",
  },
]
`;

exports[`<StudentTable/> handleGridSort() should sort the linkedlist of students correctly for direction 'DESC': firstNameDESC 1`] = `
Array [
  Student {
    "firstName": "Xerxes",
    "id": 2,
    "lastName": "Smith",
    "universityMajor": "Mathematics",
  },
  Student {
    "firstName": "George",
    "id": 3,
    "lastName": "Michael",
    "universityMajor": "History",
  },
  Student {
    "firstName": "Anton",
    "id": 1,
    "lastName": "Meyer",
    "universityMajor": "Computer Science",
  },
]
`;

exports[`<StudentTable/> handleGridSort() should sort the linkedlist of students correctly for direction 'NONE': firstNameNONE 1`] = `
Array [
  Student {
    "firstName": "Anton",
    "id": 1,
    "lastName": "Meyer",
    "universityMajor": "Computer Science",
  },
  Student {
    "firstName": "Xerxes",
    "id": 2,
    "lastName": "Smith",
    "universityMajor": "Mathematics",
  },
  Student {
    "firstName": "George",
    "id": 3,
    "lastName": "Michael",
    "universityMajor": "History",
  },
]
`;

exports[`<StudentTable/> should render correctly 1`] = `
<div>
  <nav
    className="navbar navbar-dark bg-dark justify-content-between mb-3"
  >
    <div>
      <a
        className="navbar-brand"
        href="/"
        id="navbar-logo"
      >
        Linked List & Students
      </a>
      <span
        className="navbar-text"
      >
        Sam Bokai for Algorithms & Data Structures
      </span>
    </div>
    <form
      className="form-inline"
    >
      <div
        className="form-group"
      >
        <input
          className="form-control"
          defaultValue={50}
          max={100}
          min={1}
          size={4}
          type="number"
        />
        <button
          className="btn btn-primary ml-2"
          id="new-note-button"
          onClick={[Function]}
          type="button"
        >
          Generate Students
        </button>
      </div>
      <button
        className="btn btn-danger ml-2"
        id="delete-note-button"
        onClick={[Function]}
        type="button"
      >
        Clear Table
      </button>
    </form>
  </nav>
  <ReactDataGrid
    cellNavigationMode="none"
    columnEquality={[Function]}
    columns={
      Array [
        Object {
          "filterable": true,
          "key": "id",
          "name": "ID",
          "sortable": true,
          "width": 100,
        },
        Object {
          "filterable": true,
          "key": "firstName",
          "name": "First Name",
          "sortable": true,
        },
        Object {
          "filterable": true,
          "key": "lastName",
          "name": "Last Name",
          "sortable": true,
        },
        Object {
          "filterable": true,
          "key": "universityMajor",
          "name": "Major",
          "sortable": true,
        },
      ]
    }
    enableCellAutoFocus={true}
    enableCellSelect={false}
    enableRowSelect={false}
    getCellActions={[Function]}
    headerFiltersHeight={45}
    minColumnWidth={80}
    minHeight={600}
    onAddFilter={[Function]}
    onBeforeEdit={[Function]}
    onClearFilters={[Function]}
    onGridSort={[Function]}
    overScan={
      Object {
        "colsEnd": 5,
        "colsStart": 5,
        "rowsEnd": 5,
        "rowsStart": 5,
      }
    }
    rowGetter={[Function]}
    rowHeight={35}
    rowKey="id"
    rowScrollTimeout={0}
    rowsCount={50}
    scrollToRowIndex={0}
    tabIndex={-1}
    toolbar={
      <Toolbar
        addRowButtonText="Add Row"
        enableAddRow={true}
        enableFilter={true}
        filterRowsButtonText="Filter Rows"
      >
        <button
          className="btn ml-2"
          onClick={[Function]}
        >
          Prepend New Student
        </button>
        <button
          className="btn ml-2"
          onClick={[Function]}
        >
          Append New Student
        </button>
        <div
          className="d-inline-block ml-3"
          id="student-count"
        >
          Total Students: 
          50
        </div>
      </Toolbar>
    }
  />
</div>
`;
